Packages/vcs/Lib/vcsvtk/meshfillpipeline.py

```diff
@@ -0,0 +1,305 @@
+                  % (len(self._contourLevels), len(self._contourColors),
+                     self._gm.name, self._gm.g_name))
+        for i, l in enumerate(self._contourLevels):
+            if i == 0:
+                C = [self._contourColors[i],]
+                if numpy.allclose(self._contourLevels[0][0], -1.e20):
+                    ## ok it's an extension arrow
+                    L = [self._scalarRange[0] - 1., self._contourLevels[0][1]]
+                else:
+                    L = list(self._contourLevels[i])
+                I = [indices[i],]
+            else:
+                if l[0] == L[-1] and I[-1] == indices[i]:
+                    # Ok same type lets keep going
+                    if numpy.allclose(l[1], 1.e20):
+                        L.append(self._scalarRange[1] + 1.)
+                    else:
+                        L.append(l[1])
+                    C.append(self._contourColors[i])
+                else: # ok we need new contouring
+                    tmpLevels.append(L)
+                    tmpColors.append(C)
+                    C = [self._contourColors[i],]
+                    L = self._contourLevels[i]
+                    I = [indices[i],]
+        tmpLevels.append(L)
+        tmpColors.append(C)
+
+        mappers = []
+        luts = []
+        cots = []
+        geos = []
+        for i,l in enumerate(tmpLevels):
+          # Ok here we are trying to group together levels can be, a join will
+          # happen if: next set of levels contnues where one left off AND
+          # pattern is identical
+          wholeDataMin, wholeDataMax = vcs.minmax(self._originalData1)
+          # TODO this should really just be a single polydata that is
+          # colored by scalars:
+          for j,color in enumerate(tmpColors[i]):
+              mapper = vtk.vtkPolyDataMapper()
+              lut = vtk.vtkLookupTable()
+              th = vtk.vtkThreshold()
+              th.ThresholdBetween(l[j], l[j+1])
+              th.SetInputConnection(self._vtkPolyDataFilter.GetOutputPort())
+              geoFilter2 = vtk.vtkDataSetSurfaceFilter()
+              geoFilter2.SetInputConnection(th.GetOutputPort())
+              geos.append(geoFilter2)
+              mapper.SetInputConnection(geoFilter2.GetOutputPort())
+              lut.SetNumberOfTableValues(1)
+              r, g, b = self._colorMap.index[color]
+              lut.SetTableValue(0, r/100., g/100., b/100.)
+              mapper.SetLookupTable(lut)
+              mapper.SetScalarRange(l[j], l[j+1])
+              luts.append([lut, [l[j], l[j+1], False]])
+              ## Store the mapper only if it's worth it?
+              ## Need to do it with the whole slab min/max for animation
+              ## purposes
+              if not (l[j+1] < wholeDataMin or l[j] > wholeDataMax):
+                  # TODO Charles: Why is this appending a single element list,
+                  # instead of just an object?
+                  mappers.append([mapper,])
```

Comment:
Agreed!

BTW, there is a common baseline class, Pipeline2D, and while there are a lot of _similar_ portions, many had tons of conditionals to implement different behaviors for the different backends. Pipeline2D has some of the more easily abstracted bits taken out already, and as the refactoring progresses I expect to put more of them in the base class.